DataDog/datadog-agent
# Stream logs agent logs for troubleshooting

### What does this PR do?

Adds a new agent command to stream the logs being processed by the agent. This will be useful for diagnosing issues with processing logs by enabling the ability to observe the logs in flight (just before they are sent to Datadog). 

Currently this supports 3 filters:
1. Name
2. Type
3. Source

If multiple filters are supplied it will match messages that intersect all enabled filters. (ex: name and type will only show logs that match both)

The streamed logs include all the context that would otherwise be sent to Datadog (source, tags, type, etc). The streaming is done by continuously writing and flushing to the socket initiated by the HTTPS connection as messages are made available. If the client kills the connection, messages will stop being processed and and the buffer will be flushed. The socket read/write timeout's are overridden for this endpoint so logs can be streamed as long as needed. 

### Motivation

Improve diagnosing and debugging the logs agent. 

### Describe your test plan

UT's
Tested high throughput with `flog`.

QA steps:
set up a couple of log sources (files and containers). 
run the agent and ensure some logs are flowing. 
while the agent is running run `agent stream-logs` and observe the output. Also try running with the filters.
